examples/src/bin/unbuffered-client.rs

```diff
@@ -0,0 +1,281 @@
+            "did not get a HTTP response within {MAX_ITERATIONS} iterations"
+        );
+    }
+
+    assert!(sent_request);
+    assert!(received_response);
+    assert_eq!(send_early_data, sent_early_data);
+    assert_eq!(0, incoming_used);
+    assert_eq!(0, outgoing_used);
+
+    Ok(())
+}
+
+fn try_or_resize_and_retry<E>(
+    mut f: impl FnMut(&mut [u8]) -> Result<usize, E>,
+    map_err: impl FnOnce(E) -> Result<InsufficientSizeError, Box<dyn Error>>,
+    outgoing_tls: &mut Vec<u8>,
+    outgoing_used: &mut usize,
+) -> Result<usize, Box<dyn Error>>
+where
+    E: Error + 'static,
+{
+    let written = match f(&mut outgoing_tls[*outgoing_used..]) {
+        Ok(written) => written,
+
+        Err(e) => {
+            let InsufficientSizeError { required_size } = map_err(e)?;
+            let new_len = *outgoing_used + required_size;
+            outgoing_tls.resize(new_len, 0);
+            eprintln!("resized `outgoing_tls` buffer to {new_len}B");
+
+            f(&mut outgoing_tls[*outgoing_used..])?
+        }
+    };
+
+    *outgoing_used += written;
+
+    Ok(written)
+}
+
+fn recv_tls(
+    sock: &mut TcpStream,
+    incoming_tls: &mut [u8],
+    incoming_used: &mut usize,
+) -> Result<(), Box<dyn Error>> {
+    let read = sock.read(&mut incoming_tls[*incoming_used..])?;
+    eprintln!("received {read}B of data");
+    *incoming_used += read;
+    Ok(())
+}
+
+fn send_tls(
+    sock: &mut TcpStream,
+    outgoing_tls: &[u8],
+    outgoing_used: &mut usize,
+) -> Result<(), Box<dyn Error>> {
+    sock.write_all(&outgoing_tls[..*outgoing_used])?;
+    eprintln!("sent {outgoing_used}B of data");
+    *outgoing_used = 0;
+    Ok(())
+}
+
+fn encrypt_http_request(
+    sent_request: &mut bool,
+    may_encrypt: &mut MayEncryptAppData<'_, ClientConnectionData>,
+    outgoing_tls: &mut [u8],
+    outgoing_used: &mut usize,
+) -> bool {
+    if !*sent_request {
+        let written = may_encrypt
+            .encrypt(&build_http_request(), &mut outgoing_tls[*outgoing_used..])
+            .expect("encrypted request does not fit in `outgoing_tls`");
+        *outgoing_used += written;
+        *sent_request = true;
+        eprintln!("queued HTTP request");
+        true
+    } else {
+        false
+    }
+}
+
+fn build_root_store() -> Result<RootCertStore, Box<dyn Error>> {
+    let mut root_store = RootCertStore::empty();
+    if let Some(path) = CERTFILE {
+        let certfile = File::open(path)?;
+        let mut reader = BufReader::new(certfile);
+        root_store.add_parsable_certificates(
+            rustls_pemfile::certs(&mut reader).collect::<Result<Vec<_>, _>>()?,
+        );
+    } else {
+        root_store.extend(
+            webpki_roots::TLS_SERVER_ROOTS
+                .iter()
+                .cloned(),
+        );
+    }
+    Ok(root_store)
+}
+
+fn build_http_request() -> Vec<u8> {
+    format!("GET / HTTP/1.1\r\nHost: {SERVER_NAME}\r\nConnection: close\r\nAccept-Encoding: identity\r\n\r\n").into_bytes()
+}
```

Comment:
I think this helper might be too small to justify being a separate fn. It's only invoked from `encrypt_http_response` and inlining it there feels like a clarity improvement.